### ISSUE TITLE
Not usable for development on Windows (7, 8.1, 10)

### ISSUE THREAD

**Issue by marm5r** [payara/Payara#445]
Redeploys fail because of:
https://java.net/jira/browse/GLASSFISH-21389

Automated test (e.g. Arquillian) are to slow, hang while tear down (undeploy), maybe related to:
https://java.net/jira/browse/GLASSFISH-21283

**Comment by smillidge** [comment]
Hi, 
o you have a test application we can look at. I would suspect a classloader leak.

**Comment by smillidge** [comment]
Can't reproduce this on windows 7 using CargoTracker https://cargotracker.java.net/ which is a pretty complex app that uses many JavaEE apis. 

Please try one of the latest unstable builds as a few ClassLoader leaks have been squashed for our customers.http://www.payara.fish/upstream_builds

**Comment by smillidge** [comment]
Also test app referred to in GLASSFISH-21283 redeploys in less than 2 seconds on latest builds

**Comment by marm5r** [comment]
https://java.net/jira/browse/GLASSFISH-21389 is still present in the latest unstable build.

I have published a modified version of my test application on GitHub:
https://github.com/marm5r/verify-gf-21389

Please call the ant script with *ant -Dfish=payara-pre* to run the test with Payara. The script will download Payara from http://payara.co.s3-website-eu-west-1.amazonaws.com/payara-prerelease.zip. You can also put payara-prerelease.zip into the root folder of the test application.

When I run the test application with *ant -Dfish=payara-pre* on my Windows machines the following error is logged on every run:
```
     [exec] Command deploy failed.
     [exec] remote failure: Error occurred during deployment: Exception while loading the app : WELD-001201: Error loadi
ng beans.xml jar:file:/D:/scratch/verify-gf-21389/webapp/target/webapp-1.0-SNAPSHOT/WEB-INF/lib/ejb-dep-1.0-SNAPSHOT.jar
!/META-INF/beans.xml. Please see server.log for more details.
```

Here is the complete stacktrace from server.log:
```
[2015-09-24T15:19:56.181+0200] [Payara 4.1] [SEVERE] [NCLS-CORE-00026] [javax.enterprise.system.core] [tid: _ThreadID=54 _ThreadName=admin-listener(2)] [timeMillis: 1443100796181] [levelValue: 1000] [[
  Exception during lifecycle processing
org.jboss.weld.exceptions.IllegalStateException: WELD-001201: Error loading beans.xml jar:file:/D:/scratch/verify-gf-21389/webapp/target/webapp-1.0-SNAPSHOT/WEB-INF/lib/ejb-dep-1.0-SNAPSHOT.jar!/META-INF/beans.xml
	at org.jboss.weld.xml.BeansXmlParser.parse(BeansXmlParser.java:124)
	at org.jboss.weld.bootstrap.WeldBootstrap.parse(WeldBootstrap.java:131)
	at org.glassfish.weld.BeanDeploymentArchiveImpl.parseBeansXML(BeanDeploymentArchiveImpl.java:755)
	at org.glassfish.weld.BeanDeploymentArchiveImpl.populate(BeanDeploymentArchiveImpl.java:465)
	at org.glassfish.weld.BeanDeploymentArchiveImpl.<init>(BeanDeploymentArchiveImpl.java:150)
	at org.glassfish.weld.RootBeanDeploymentArchive.createModuleBda(RootBeanDeploymentArchive.java:94)
	at org.glassfish.weld.RootBeanDeploymentArchive.<init>(RootBeanDeploymentArchive.java:87)
	at org.glassfish.weld.RootBeanDeploymentArchive.<init>(RootBeanDeploymentArchive.java:75)
	at org.glassfish.weld.DeploymentImpl.createModuleBda(DeploymentImpl.java:712)
	at org.glassfish.weld.DeploymentImpl.<init>(DeploymentImpl.java:131)
	at org.glassfish.weld.WeldDeployer.load(WeldDeployer.java:473)
	at org.glassfish.weld.WeldDeployer.load(WeldDeployer.java:101)
	at org.glassfish.internal.data.ModuleInfo.load(ModuleInfo.java:206)
	at org.glassfish.internal.data.ApplicationInfo.load(ApplicationInfo.java:313)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:496)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:219)
	at org.glassfish.deployment.admin.DeployCommand.execute(DeployCommand.java:490)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:539)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:535)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:360)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2.execute(CommandRunnerImpl.java:534)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:565)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:557)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:360)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:556)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:1464)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl.access$1300(CommandRunnerImpl.java:109)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1846)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1722)
	at org.glassfish.admin.rest.resources.admin.CommandResource.executeCommand(CommandResource.java:404)
	at org.glassfish.admin.rest.resources.admin.CommandResource.execCommandSimpInMultOut(CommandResource.java:234)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161)
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:160)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102)
	at org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:309)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:292)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1139)
	at org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpContainer.service(GrizzlyHttpContainer.java:375)
	at org.glassfish.admin.rest.adapter.JerseyContainerCommandService$3.service(JerseyContainerCommandService.java:173)
	at org.glassfish.admin.rest.adapter.RestAdapter.service(RestAdapter.java:179)
	at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:459)
	at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:167)
	at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:206)
	at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:180)
	at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:235)
	at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:119)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:283)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:200)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:132)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:111)
	at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:77)
	at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:536)
	at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:112)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:117)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:56)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:137)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:591)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:571)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.FileNotFoundException: JAR entry META-INF/beans.xml not found in D:\scratch\verify-gf-21389\webapp\target\webapp-1.0-SNAPSHOT\WEB-INF\lib\ejb-dep-1.0-SNAPSHOT.jar
	at sun.net.www.protocol.jar.JarURLConnection.connect(JarURLConnection.java:142)
	at sun.net.www.protocol.jar.JarURLConnection.getInputStream(JarURLConnection.java:150)
	at java.net.URL.openStream(URL.java:1038)
	at org.jboss.weld.xml.BeansXmlParser.parse(BeansXmlParser.java:101)
	... 75 more
]]
```

**Comment by marm5r** [comment]
@smillidge https://java.net/jira/browse/GLASSFISH-21389 is IMHO not reproducible with cargo-tracker because it is just a single war.

*java.io.FileNotFoundException: JAR entry META-INF/beans.xml not found* is related to WEB-INF/lib dependencies containing a META-INF/beans.xml.

To reproduce the exception manually with the test application (https://github.com/marm5r/verify-gf-21389) please follow the steps as mentioned in https://java.net/jira/browse/GLASSFISH-21389:

1. open the Maven project with NetBeans 8.0.2 (Java 8)
2. build project
3. run sub-project 'webapp' with latest Payara build
4. open sub-project ejb-dep
5. modify ejb.SessionBean, e.g. change *if (false)* to *if (true)*
6. build sub-project ejb-dep
7. run sub-project 'webapp' again to force redeploy
8. Exception WELD-001201: Error loading beans.xml... will be thrown

The exception will NOT occur if you leave out step 5!

The ant script from the test application performs the same steps  - without NetBeans of course...

**Comment by smillidge** [comment]
Thanks for the test case we'' investigate

**Comment by marm5r** [comment]
I have opened a new issue for the problem with Arquillian:
#456

**Comment by ratcashdev** [comment]
Similar finding (and test case) here: https://github.com/omnifaces/omnifaces/issues/115 likely related to https://java.net/jira/browse/GLASSFISH-21283

**Comment by smillidge** [comment]
We are doing a lot of work in this area related to various classloader leaks.

**Comment by smillidge** [comment]
Can you try on the latest Payara release as a lot of changes have been made to prevent these file locks.

**Comment by ratcashdev** [comment]
4.1.154 still exhibits the problem with https://github.com/omnifaces/omnifaces/issues/115.

**Comment by smillidge** [comment]
I can't reproduce any problem with omnifaces on Linux or Windows. Application with Omnifaces redeploys fine on both platforms without delays and there are no omnifaces object instances in the heap after undeploy. This is on the latest snapshot Payara Builds

**Comment by ratcashdev** [comment]
My test case:

sample pom.xml
```
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <groupId>com.ratcash</groupId>
    <artifactId>RedeployTest</artifactId>
    <version>1.0-SNAPSHOT</version>
    <packaging>war</packaging>
    <dependencies>
        <dependency>
            <groupId>org.omnifaces</groupId>
            <artifactId>omnifaces</artifactId>
            <version>2.2</version>
        </dependency>
        <dependency>
            <groupId>javax</groupId>
            <artifactId>javaee-api</artifactId>
            <version>7.0</version>
            <scope>provided</scope>
        </dependency>
        
    </dependencies>
    <build>
        <finalName>RedeployTest</finalName>
    </build>
    <properties>
        <maven.compiler.source>1.8</maven.compiler.source>
        <maven.compiler.target>1.8</maven.compiler.target>
        <failOnMissingWebXml>false</failOnMissingWebXml>
    </properties>
</project>
```

Sample JAVA file:
```
package com.airhacks;

import javax.ws.rs.ApplicationPath;
import javax.ws.rs.core.Application;

/**
 * Configures a JAX-RS endpoint. Delete this class, if you are not exposing
 * JAX-RS resources in your application.
 *
 * @author airhacks.com
 */
@ApplicationPath("api")
public class JAXRSConfiguration extends Application {

}
```


**Scenario 1:**
Compile, package deploy inside Netbeans (using 8.1 on Windows 7, 64 bit).
After deploying the project from Netbeans right-click the project and select "Clean and Build":
```
--- maven-clean-plugin:2.4.1:clean (default-clean) @ RedeployTest ---
Deleting C:\Path\NetBeansProjects\RedeployTest\target
------------------------------------------------------------------------
BUILD FAILURE
------------------------------------------------------------------------
Total time: 0.574s
Finished at: Mon Dec 14 15:40:38 CET 2015
Final Memory: 6M/304M
------------------------------------------------------------------------
Failed to execute goal org.apache.maven.plugins:maven-clean-plugin:2.4.1:clean (default-clean) on project RedeployTest: Failed to clean project: Failed to delete C:\Path\NetBeansProjects\RedeployTest\target\RedeployTest\WEB-INF\lib\omnifaces-2.2.jar -> [Help 1]
```

**Scenario 2:**
1. Build the .war via MAVEN. 
2. Copy the resulting WAR file to the autodeploy folder of Payara manually.
3. Wait for successful deployment
4. delete the war file

You'll notice the the following file appear in the applications/RedeployTest folder:
.glassfishStaleFiles with the content:
```
WEB-INF/
WEB-INF/lib/
WEB-INF/lib/omnifaces-2.2.jar
```
I know this is not OMNIFACE's problem, I just used their lib to manifest the issue.

**Comment by ratcashdev** [comment]
Test on Payara 4.1.161-SNAPSHOT Build 33 on Windws 7 64bit with JDK 1.8_60:
Scenario 1: fails
Scenario 2: fails

**Comment by ratcashdev** [comment]
Test on Payara 4.1.161, Windws 7 64bit with JDK 1.8_60:
Scenario 1: Fails
Scenario 2: fails

**Comment by smillidge** [comment]
This is because Payara Server keeps a file descriptor open after deployment to the application. Windows doesn't let you delete files with open file descriptors unlike linux. Deployment via asadmin or admin console should work fine.

**Comment by ratcashdev** [comment]
I'd assume that no file descriptors would remain open after a proper undeployment. Why would there be an open file descriptor (opened by Glassfish), if Glassfish knows it has to close the descriptor and undeploy the app?

**Comment by smillidge** [comment]
No file descriptors are left open after undeployment if you use asadmin undeploy. It's a bug if they are left open. We have tracked down and squashed many of these bugs. However file descriptors are open while the application is active.

**Comment by smillidge** [comment]
Your scenarios as described involve deleting the war while deployed. I don't think that will be possible on Windows

**Comment by ratcashdev** [comment]
AFAIK, scenario 2 is the use case for auto-deployment and undeployment. At least according to the logs they are treated as such. Keep in mind that the only file not deleted is the omnifaces.jar. That file is not deleted because it contains a beans.xml. If it didn't, there would be no problem with the file descriptor.

**Comment by ratcashdev** [comment]
Otherwise, yes, deploying/undeploying from the admin GUI is OK, no artefacts/jars were left behind, regardless of whether the jars contain a beans.xml or not.

**Comment by ratcashdev** [comment]
Anyway, apparently, this was a known-issue to the Glassfish-devs. It seems like they chose to implement a workaround, i.e. create the .glassfishStaleFiles file. Moreover, while removing such jars is not possible, Glassfish still allows overwriting them (i.e. redeploy).

**Comment by smillidge** [comment]
I can give you a patched jar if you would like to test against the scenarios above. it works for me for your scenarios above. The patch is based on #699

**Comment by smillidge** [comment]
Tested with Netbeans and #699 

Scenario 1 still fails, however if you force a GC on the server, via jvisualvm, then you can clean and build again. Or if you undeploy via NetBeans or if you undeploy via the admin console there is no problem with the clean and build cycle.

On the other hand all the NetBeans deploy on save functionality seems to work fine so a clean and build isn't necessary for rapid development. 

We'll still investigate why scenario 1 is failing. 

Scenario 2 works fine.

**Comment by ratcashdev** [comment]
I would gladly test the patched jar!

**Comment by smillidge** [comment]
This should also fix Scenario 1.

The attached zip should be renamed to deployment-admin.jar and dropped into your glassfish modules directory and then restart the server.

[deployment-admin.zip](https://github.com/payara/Payara/files/170860/deployment-admin.zip)

**Comment by ratcashdev** [comment]
Successful test results as well. 
I was curious to see whether this forced undeployment would break versioned deployments, like 
`asadmin deploy --name="redeploy:2" --enabled=false RedeployTest.war` but I am happy to report that it works as expected.

**Comment by smillidge** [comment]
Fantastic news thanks for testing.

**Comment by smillidge** [comment]
Merged #699
